File: src/com/anthonymandra/rawdroid/ImageViewActivity.java
package com.anthonymandra.rawdroid;

import com.actionbarsherlock.view.Menu;
import com.actionbarsherlock.widget.ShareActionProvider.OnShareTargetSelectedListener;
import com.android.gallery3d.app.PhotoDataAdapter.DataListener;
import com.android.gallery3d.app.PhotoPage;
import com.android.gallery3d.data.ContentListener;
import com.android.gallery3d.data.MediaItem;
import com.android.gallery3d.data.MediaObject;
import com.anthonymandra.framework.LicenseManager;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.widget.Toast;

import java.io.File;
import java.util.WeakHashMap;

public class ImageViewActivity extends PhotoPage implements DataListener
{
	private static final String TAG = ImageViewActivity.class.getSimpleName();

	// private DecodeRawTask decodeTask;
	// private FrameLayout decodeProgress;

    @Override
	public void onCreate(Bundle savedInstanceState)
	{
    	mContentView = R.layout.viewer_layout;
		super.onCreate(savedInstanceState);

        initialize();

		SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(this);
        settings.registerOnSharedPreferenceChangeListener(this);
	}
    
	@Override
	protected void updateAfterDelete()
	{
		updateImageSource();
	}

    @Override
	protected void updateAfterRestore()
	{
		updateImageSource();
	}

	private void updateImageSource()
	{
		updateViewerItems();
		notifyContentChanged();
	}

	@Override
	public void onSingleTapConfirmed()
	{
		togglePanels();
	}

	@Override
	/**
	 * This seems to occur when the first image is loaded.
	 */
	public void onCurrentImageUpdated()
	{
        super.onCurrentImageUpdated();
        setShareUri(getCurrentItem().getSwapUri());
		updateImageDetails();
	}

	@Override
	/**
	 * This occurs whenever the current image changes.
	 */
	public void onPhotoChanged(int index, Uri path)
	{
        super.onPhotoChanged(index, path);
		updateImageDetails();
	}

    private WeakHashMap<ContentListener, Object> mListeners =
            new WeakHashMap<ContentListener, Object>();

    // NOTE: The MediaSet only keeps a weak reference to the listener. The
    // listener is automatically removed when there is no other reference to
    // the listener.
    public void addContentListener(ContentListener listener) {
        mListeners.put(listener, null);
    }

    public void removeContentListener(ContentListener listener) {
        mListeners.remove(listener);
    }

    // This should be called by subclasses when the content is changed.
    public void notifyContentChanged() {
        for (ContentListener listener : mListeners.keySet()) {
            listener.onContentDirty();
        }
    }

    @Override
    public MediaItem getCurrentItem() {
        return mModel.getCurrentItem();
    }

    @Override
    public Bitmap getCurrentBitmap() {
        return mModel.getCurrentBitmap();
    }

    @Override
    public void goToPrevPicture() {
        mPhotoView.goToPrevPicture();
    }

    @Override
    public void goToNextPicture() {
        mPhotoView.goToNextPicture();
    }

    @Override
    public void goToFirstPicture() {
        mModel.moveTo(0);
    }

    @Override
    public LicenseManager getLicenseManager() {
        return mLicenseManager;
    }
}

File: app/src/main/java/com/anthonymandra/rawdroid/ImageViewActivity.kt
package com.anthonymandra.rawdroid

import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.preference.PreferenceManager
import com.android.gallery3d.app.PhotoPage
import com.android.gallery3d.data.ContentListener
import com.anthonymandra.framework.License
import java.util.*

class ImageViewActivity : PhotoPage() {

    public override val contentView: Int
        get() = R.layout.viewer_twopane

    private val mListeners = WeakHashMap<ContentListener, Any>()

    override val currentItem: Uri?
        get() = mModel.currentItem

    override val currentBitmap: Bitmap
        get() = mModel.currentBitmap

    override fun updateMessage(message: String?) {

    }

    override fun setMaxProgress(max: Int) {

    }

    override fun incrementProgress() {

    }

    override fun endProgress() {

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val settings = PreferenceManager.getDefaultSharedPreferences(this)
        settings.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onImageSetChanged() {
        // TODO: This should be handled in ViewerActivity and requery
        mPhotoView.setUpdateForContentChange(true)
        notifyContentChanged()
    }

    override fun onSingleTapConfirmed() {
        togglePanels()
    }

    override fun onCurrentImageUpdated() {
        super.onCurrentImageUpdated()
        if (mRequiresHistogramUpdate)
            updateHistogram(currentBitmap)
    }

    override fun onCommitDeleteImage(toDelete: Uri) {
        deleteImage(toDelete)
    }

    // NOTE: The MediaSet only keeps a weak reference to the listener. The
    // listener is automatically removed when there is no other reference to
    // the listener.
    override fun addContentListener(listener: ContentListener) {
        mListeners[listener] = null
    }

    override fun removeContentListener(listener: ContentListener) {
        mListeners.remove(listener)
    }

    // This should be called by subclasses when the content is changed.
    protected fun notifyContentChanged() {
        for (listener in mListeners.keys) {
            listener.onContentDirty()
        }
    }

    override fun goToPrevPicture() {
        mPhotoView.goToPrevPicture()
    }

    override fun goToNextPicture() {
        mPhotoView.goToNextPicture()
    }

    override fun setLicenseState(state: License.LicenseState) {
        super.setLicenseState(state)
        mPhotoView.setLicenseState(state)
    }

    companion object {

        private val TAG = ImageViewActivity::class.java.simpleName
    }
}